crates/sui-graphql-rpc/src/types/move_package.rs

```diff
@@ -569,6 +596,152 @@ impl MovePackage {
             Error::Internal(format!("{address} is not a package"))
         })?))
     }
+
+    /// Query the database for a `page` of Move packages. The Page uses the checkpoint sequence
+    /// number the package was created at, its original ID, and its version as the cursor. The query
+    /// can optionally be filtered by a bound on the checkpoints the packages were created in.
+    ///
+    /// The `checkpoint_viewed_at` parameter represents the checkpoint sequence number at which this
+    /// page was queried. Each entity returned in the connection will inherit this checkpoint, so
+    /// that when viewing that entity's state, it will be as if it is being viewed at this
+    /// checkpoint.
+    ///
+    /// The cursors in `page` may also include checkpoint viewed at fields. If these are set, they
+    /// take precedence over the checkpoint that pagination is being conducted in.
+    pub(crate) async fn paginate_by_checkpoint(
+        db: &Db,
+        page: Page<Cursor>,
+        after_checkpoint: Option<u64>,
+        before_checkpoint: Option<u64>,
+        checkpoint_viewed_at: u64,
+    ) -> Result<Connection<String, MovePackage>, Error> {
+        let cursor_viewed_at = page.validate_cursor_consistency()?;
+        let checkpoint_viewed_at = cursor_viewed_at.unwrap_or(checkpoint_viewed_at);
+
+        // Clamp the "before checkpoint" bound by "checkpoint viewed at".
+        let before_checkpoint =
+            (checkpoint_viewed_at + 1).min(before_checkpoint.unwrap_or(u64::MAX));
+
+        let (prev, next, results) = db
+            .execute(move |conn| {
+                let mut q = query!(
+                    r#"
+                    SELECT
+                            p.original_id,
+                            o.*
+                    FROM
+                            packages p
+                    INNER JOIN
+                            objects_history o
+                    ON
+                            p.package_id = o.object_id
+                    AND     p.package_version = o.object_version
+                    AND     p.checkpoint_sequence_number = o.checkpoint_sequence_number
```

Comment:
oh no.. did we give up on using the dsl haha

was it the typings or something else that gave you trouble

Reply:
This was indeed a type issue. Specifically, the way `paginate_query` is written, it expects the underlying query's source to be the table that the cursor is paginating from, but in this case, the source is a join of two tables, and that messes everything up.